### PR TITLE
set_default_section: Handle casing properly

### PR DESCRIPTION
**Pull request by sils**
This commit was coauthored by Maximilian Scholz.

Fixes https://github.com/coala/coala/issues/4182

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/settings/Section.py [hunk 1]
```diff
@@ -315,12 +315,23 @@ def set_default_section(self, sections, section_name=None):
         >>> section.defaults.name
         'all'
 
+        This works for uppercase sections as well. The key of the sections dict
+        is expected to be lowered though!
+
+        >>> sections = {'c': Section('C')}
+        >>> section = Section('C.something')
+        >>> section.set_default_section(sections)
+        >>> section.defaults.name
```
Comment (Udayan12167) [inline coalib/settings/Section.py]: Can you add a test for the section name `c.Something.Else`

Comment (Mixih) [inline coalib/settings/Section.py]: Why? We don't support multiple inheritance do we?

File: coalib/settings/Section.py [hunk 2]
```diff
@@ -315,12 +315,23 @@ def set_default_section(self, sections, section_name=None):
...
+
         :param sections:     A dictionary of sections.
         :param section_name: Optional section name argument to find the default
                              section for. If not given then use member section
                              name.
         """
-        default_section = '.'.join((section_name or self.name).split('.')[:-1])
+        default_section = '.'.join(
+            (section_name or self.name).split('.')[:-1]
+        ).lower()
```
Comment (userzimmermann) [inline coalib/settings/Section.py]: Can be simplified to:

```python
default_section = (section_name or self.name).rsplit('.', 1)[0].lower()
```

File: coalib/settings/Section.py (hunk 1 above)
Comment (Mixih) [inline coalib/settings/Section.py]: NVM, I see that this would be a good idea.

**Comment by jayvdb** [comment]
All builds are failing due to a bug.

```python
File "/home/travis/build/coala/coala/coalib/settings/Section.py", line 215, in __getitem__
    return self.defaults[key]
  File "/home/travis/build/coala/coala/coalib/settings/Section.py", line 201, in __getitem__
    key = self.__prepare_key(item)
  File "/home/travis/build/coala/coala/coalib/settings/Section.py", line 93, in __prepare_key
    return str(key).lower().strip()
RuntimeError: maximum recursion depth exceeded in comparison
```

**Comment by sils** [comment]
ok I'll be reverting @userzimmermann s simplification since it seems to break things

File: coalib/settings/Section.py (hunk 2 above)
Comment (jayvdb) [inline coalib/settings/Section.py]: apparently that caused failures.

Comment (userzimmermann) [inline coalib/settings/Section.py]: @sils @jayvdb Oops! Right! It doesn't properly cover the case that no `'.'` is included in the string :D

**Comment by jayvdb** [comment]
ack 9ebfceb

**Comment by sils** [comment]
@rultor merge

**Comment by sils** [comment]
@rultor merge

**Comment by Mixih** [comment]
backporting to 0.11 now!

**Pull request merged**